relay_messages_test.go

```diff
@@ -352,6 +364,113 @@ func TestLazyCallArg2Offset(t *testing.T) {
 	})
 }
 
+func TestLazyCallReqSetTChanThriftArg2(t *testing.T) {
+	tests := []struct {
+		msg            string
+		bufKV          map[string]string
+		wantKV         map[string]string // if not set, use bufKV
+		rawArgScheme   bool
+		overrideBufLen int
+		wantBadErr     string
+	}{
+		{
+			msg: "two key value pairs",
+			bufKV: map[string]string{
+				"key":  "val",
+				"key2": "val2",
+			},
+		},
+		{
+			msg: "length not enough to cover key len",
+			bufKV: map[string]string{
+				"key": "val",
+			},
+			overrideBufLen: 3, // 2 (nh) + 2 - 1
+			wantBadErr:     "invalid key offset 2 (arg2 len 3)",
+		},
+		{
+			msg: "length not enough to cover key",
+			bufKV: map[string]string{
+				"key": "val",
+			},
+			overrideBufLen: 6, // 2 (nh) + 2 + len(key) - 1
+			wantBadErr:     "invalid value offset 7 (key offset 4, key len 3, arg2 len 6)",
+		},
+		{
+			msg: "length not enough to cover value len",
+			bufKV: map[string]string{
+				"key": "val",
+			},
+			overrideBufLen: 8, // 2 (nh) + 2 + len(key) + 2 - 1
+			wantBadErr:     "invalid value offset 7 (key offset 4, key len 3, arg2 len 8)",
+		},
+		{
+			msg: "length not enough to cover value",
+			bufKV: map[string]string{
+				"key": "val",
+			},
+			overrideBufLen: 10, // 2 (nh) + 2 + len(key) + 2 + len(val) - 2
+			wantBadErr:     "value exceeds arg2 range (offset 9, len 3, arg2 len 10)",
+		},
+		{
+			msg:   "no key value pairs",
+			bufKV: map[string]string{},
+		},
+		{
+			msg:          "not tchannel thrift",
+			bufKV:        map[string]string{"key": "val"},
+			rawArgScheme: true,
+			wantBadErr:   "non thrift scheme",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.msg, func(t *testing.T) {
+			withLazyCallReqCombinations(func(crt testCallReq) {
+				arg2Buf := thriftarg2test.BuildKVBuffer(tt.bufKV)
+				if tt.overrideBufLen > 0 {
+					arg2Buf = arg2Buf[:tt.overrideBufLen]
+				}
+				cr := crt.reqWithParams(testCallReqParams{
+					hasTChanThrift: !tt.rawArgScheme,
+					arg2Buf:        arg2Buf,
+				})
+				gotIter := make(map[string]string)
+				iter, err := cr.Arg2Iterator()
+				for err == nil {
+					gotIter[string(iter.Key())] = string(iter.Value())
+					iter, err = iter.Next()
+				}
+				if tt.wantBadErr != "" {
+					require.NotEqual(t, io.EOF, err)
+					assert.Contains(t, err.Error(), tt.wantBadErr)
+				} else {
+					assert.Equal(t, io.EOF, err)
+					wantKV := tt.wantKV
+					if wantKV == nil {
+						wantKV = tt.bufKV
+					}
+					assert.Equal(t, wantKV, gotIter, "%v", crt)
```

Comment:
do you want to do `%+v` to make the frame more readable, and maybe add context. E.g., "unexpected arg2 key values for call request: %+v"

that said, the crt isn't really easy to display (since it'll be lots of bytes), so might not be worth having it at all